app/api/route/request.py

```diff
@@ -1,27 +1,39 @@
+from dataclasses import fields
 from typing import Any, Iterable
 
-from pydantic import BaseModel
+from marshmallow_dataclass import dataclass as marshmallow_dataclass
 
+import api.logging
+from api.route.api_context import ApiContext
+from api.route.models.base_api_model import BaseApiModel
 
-class BaseRequestModel(BaseModel):
-    class Config:
-        orm_mode = True
+logger = api.logging.get_logger(__name__)
 
-    def get_set_params(self) -> Iterable[tuple[str, Any]]:
+PARAM_UNSET_IN_REQUEST: str = "PARAM_UNSET_IN_REQUEST"
+
+
+@marshmallow_dataclass
+class BaseRequestModel(BaseApiModel):
+    def get_set_params(self, api_context: ApiContext) -> Iterable[tuple[str, Any]]:
```

Comment:
This is easily the most frustrating thing I ran into. For an update function, we ONLY want to set parameters that the user passed in. So if you do `PATCH` and only set `first_name` we should only work with that. Pydantic was useful in that it kept track of the values passed into its init, but dataclasses don't do that. For dataclasses, we have two options:

1. In the request, we can set `UserPatchParams.Schema(partial=True)` which will call the dataclass with only the params passed in. But for this we need default values for every init variable so the dataclasses init function gets generated as `def __init__(self, param1 = <default>, param2 = <default>, ...)`. I tried using this approach and forcing all the defaults to be `PARAM_UNSET_IN_REQUEST` but this breaks the swagger endpoint for any non-strings during serialization. So this approach doesn't work
2. We allow the dataclass to be initialized with values defaulting to `None` and then refer back to the raw API request object from Flask to see what was actually passed in. This works, but any nested params aren't going to work well here. Although that might just make you split your patch methods up as is recommended.

Open to any alternative ideas.

Reply:
There's two ways to use Marshmallow, as a schema, or as dataclasses which "wrap" the schema.

If you use just the straight Schema approach, the request body objects you work with are dictionaries/JSON. They've been validated, and types will be converted (eg. a date string to a date object), but you'll have more typing headaches. 

With marshmallow-dataclasses, it does the same Schema validation as before, but then calls your dataclass to actually give you something better to work with. It's effectively the same as doing:
```py
dataclass_obj = YourDataclass(**request_json)
```
at the start of each of your request methods, but handling nested objects a bit better.

You can tell it to use `partial=True` with marshmallow dataclasses, BUT if you do that without any defaults, you're calling the `__init__` function of your dataclass with only a few required parameters, which will error. If you give those parameters defaults (eg. `__init__(self, x = None, y = None)`) then you can't tell what was actually set, and what wasn't. I tried various ways to set defaults that made it obvious what was set, but then you need to handle every separate type (for strings putting `THIS_WAS_NOT_SET` works, but what do you do for dates - the value has to be valid).

Reply:
You might be right. I'd like to keep dataclasses, but let's see what the effort would be to have the endpoints based purely on the schema/json, even if it means we need to do `MyDataclass(**request_json)` at the start of every request then so be it if it simplifies the patch approach elsewhere.

Reply:
sounds good!

Reply:
I can't seem to figure out a great way to do that unfortunately. If I define just dataclasses and then convert them, it doesn't handle nested objects/inheritance very well without several workarounds there.

I might be able to inject the info into the dataclass while it loads, but even that is proving to be a bit hacky.

The "clean" way to do it is to define a Schema, and separately define a dataclass, but then we're duplicating every model, and I can't figure out a way to dedupe these.

Reply:
not handling inheritance should be ok, i think we can make do without inheritance for data models.

not handling nested objects seems like a bigger issue. could you elaborate on what you mean by that? if you have classes like

```
@dataclass
class UserRole:
  ...

@dataclass
class User:
  name: str
  role: UserRole
```

then do `UserSchema = class_schema(User)`, what happens in that situation? apologies if i'm not understanding the issue properly.

Reply:
```py
@dataclass
class UserRole:
  description: str

@dataclass
class User:
  name: str
  roles: list[UserRole]
```
When you use this as the basis for the Schema, it doesn't add the UserRole to the schema for some reason and calling with `{"name":"bob", "roles": [{..anything.}]}` errors as the field is unknown to it.

Reply:
oh. that's odd, the example on [their docs](https://pypi.org/project/marshmallow-dataclass/) seems to handle that use case:

```python
from dataclasses import dataclass, field
from typing import List, Optional

import marshmallow_dataclass
import marshmallow.validate


@dataclass
class Building:
    # field metadata is used to instantiate the marshmallow field
    height: float = field(metadata={"validate": marshmallow.validate.Range(min=0)})
    name: str = field(default="anonymous")


@dataclass
class City:
    name: Optional[str]
    buildings: List[Building] = field(default_factory=list)


city_schema = marshmallow_dataclass.class_schema(City)()

city = city_schema.load(
    {"name": "Paris", "buildings": [{"name": "Eiffel Tower", "height": 324}]}
)
# => City(name='Paris', buildings=[Building(height=324.0, name='Eiffel Tower')])

city_dict = city_schema.dump(city)
# => {'name': 'Paris', 'buildings': [{'name': 'Eiffel Tower', 'height': 324.0}]}
```

i wonder if it's because we need the `field(default_factory=list)` ?

Reply:
Maybe? We need to specify `field(...)` otherwise when the dataclass generates, it complains that you've specified default and non-default `__init__` params out of order. Strangely, the `@marshmallow_dataclass` decorator handles this better and knows if we don't specify something, to properly recurse and use the whole nested class.